shuttle/src/sync/mod.rs

```diff
@@ -21,7 +22,136 @@ pub use rwlock::RwLock;
+    }
+
+    pub(crate) const fn new(
+        resource_type: ResourceType,
+        static_create_location: &'static Location<'static>,
+        parent_task_signature: u64,
+        create_location_counter: u32,
+    ) -> Self {
+        let mut hasher = SipHasher::new();
+        let file: &'static str = static_create_location.file();
+        hasher.hash(file.as_bytes());
+        hasher.write_u32(static_create_location.line());
+        hasher.write_u32(static_create_location.column());
+        let static_create_location_hash = hasher.finish();
+        hasher.write_u64(static_create_location_hash);
+        hasher.write_u64(parent_task_signature);
+        hasher.write_u32(create_location_counter);
+        let signature_hash = hasher.finish();
+
+        Self {
+            resource_type,
+            static_create_location,
+            parent_task_signature,
+            static_create_location_hash,
+            create_location_counter,
+            signature_hash,
+        }
+    }
+
+    /// Hash of the static location within the source code where the task was spawned
+    pub(crate) fn static_create_location_hash(&self) -> u64 {
+        self.static_create_location_hash
+    }
+
+    /// Combined signature of the static location and dynamic context
+    /// context where the task was spawned.
+    pub(crate) fn signature_hash(&self) -> u64 {
+        self.signature_hash
+    }
+}
+
+impl Debug for ResourceSignature {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("ResourceSignature")
+            .field("type", &self.resource_type)
+            .field("static_create_location", &self.static_create_location)
+            .field("parent_task_signature", &self.parent_task_signature)
+            .field("create_location_counter", &self.create_location_counter)
+            .finish()
+    }
+}
+
+impl Display for ResourceSignature {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let suffix = match self.create_location_counter {
+            1 => "st",
+            2 => "nd",
+            3 => "rd",
+            _ => "th",
+        };
+        write!(
+            f,
+            "{:?}[ {:?}{}@{}:{:?}:{:?} on {} ]",
+            self.resource_type,
+            self.create_location_counter,
+            suffix,
+            self.static_create_location.file(),
+            self.static_create_location.line(),
+            self.static_create_location.column(),
+            self.parent_task_signature,
+        )
+    }
+}
+
+impl Hash for ResourceSignature {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.static_create_location_hash().hash(state);
+        self.parent_task_signature.hash(state);
+        self.create_location_counter.hash(state);
+    }
+}
+
+impl PartialEq for ResourceSignature {
+    fn eq(&self, other: &Self) -> bool {
+        self.signature_hash() == other.signature_hash()
```

Comment:
So we consider the hash as the "identity" of the signature, and don't care about collisions? I guess there is not much point since the parent info is just a hash.